### PR TITLE
[SPARK-30027][SQL] Support codegen for aggregate filters in HashAggregateExec

### PR DESCRIPTION
**Pull request by maropu**
### What changes were proposed in this pull request?

This pr intends to support code generation for `HashAggregateExec` with filters.

Quick benchmark results:
```
$ ./bin/spark-shell --master=local[1] --conf spark.driver.memory=8g --conf spark.sql.shuffle.partitions=1 -v

scala> spark.range(100000000).selectExpr("id % 3 as k1", "id % 5 as k2", "rand() as v1", "rand() as v2").write.saveAsTable("t")
scala> sql("SELECT k1, k2, AVG(v1) FILTER (WHERE v2 > 0.5) FROM t GROUP BY k1, k2").write.format("noop").mode("overwrite").save()

>> Before this PR
Elapsed time: 16.170697619s 

>> After this PR
Elapsed time: 6.7825313s  
```

The query above is compiled into code below;

```
...
/* 285 */   private void agg_doAggregate_avg_0(boolean agg_exprIsNull_2_0, org.apache.spark.sql.catalyst.InternalRow agg_unsafeRowAggBuffer_0, double agg_expr_2_0) throws java.io.IOException {
/* 286 */     // evaluate aggregate function for avg
/* 287 */     boolean agg_isNull_10 = true;
/* 288 */     double agg_value_12 = -1.0;
/* 289 */     boolean agg_isNull_11 = agg_unsafeRowAggBuffer_0.isNullAt(0);
/* 290 */     double agg_value_13 = agg_isNull_11 ?
/* 291 */     -1.0 : (agg_unsafeRowAggBuffer_0.getDouble(0));
/* 292 */     if (!agg_isNull_11) {
/* 293 */       agg_agg_isNull_12_0 = true;
/* 294 */       double agg_value_14 = -1.0;
/* 295 */       do {
/* 296 */         if (!agg_exprIsNull_2_0) {
/* 297 */           agg_agg_isNull_12_0 = false;
/* 298 */           agg_value_14 = agg_expr_2_0;
/* 299 */           continue;
/* 300 */         }
/* 301 */
/* 302 */         if (!false) {
/* 303 */           agg_agg_isNull_12_0 = false;
/* 304 */           agg_value_14 = 0.0D;
/* 305 */           continue;
/* 306 */         }
/* 307 */
/* 308 */       } while (false);
/* 309 */
/* 310 */       agg_isNull_10 = false; // resultCode could change nullability.
/* 311 */
/* 312 */       agg_value_12 = agg_value_13 + agg_value_14;
/* 313 */
/* 314 */     }
/* 315 */     boolean agg_isNull_15 = false;
/* 316 */     long agg_value_17 = -1L;
/* 317 */     if (!false && agg_exprIsNull_2_0) {
/* 318 */       boolean agg_isNull_18 = agg_unsafeRowAggBuffer_0.isNullAt(1);
/* 319 */       long agg_value_20 = agg_isNull_18 ?
/* 320 */       -1L : (agg_unsafeRowAggBuffer_0.getLong(1));
/* 321 */       agg_isNull_15 = agg_isNull_18;
/* 322 */       agg_value_17 = agg_value_20;
/* 323 */     } else {
/* 324 */       boolean agg_isNull_19 = true;
/* 325 */       long agg_value_21 = -1L;
/* 326 */       boolean agg_isNull_20 = agg_unsafeRowAggBuffer_0.isNullAt(1);
/* 327 */       long agg_value_22 = agg_isNull_20 ?
/* 328 */       -1L : (agg_unsafeRowAggBuffer_0.getLong(1));
/* 329 */       if (!agg_isNull_20) {
/* 330 */         agg_isNull_19 = false; // resultCode could change nullability.
/* 331 */
/* 332 */         agg_value_21 = agg_value_22 + 1L;
/* 333 */
/* 334 */       }
/* 335 */       agg_isNull_15 = agg_isNull_19;
/* 336 */       agg_value_17 = agg_value_21;
/* 337 */     }
/* 338 */     // update unsafe row buffer
/* 339 */     if (!agg_isNull_10) {
/* 340 */       agg_unsafeRowAggBuffer_0.setDouble(0, agg_value_12);
/* 341 */     } else {
/* 342 */       agg_unsafeRowAggBuffer_0.setNullAt(0);
/* 343 */     }
/* 344 */
/* 345 */     if (!agg_isNull_15) {
/* 346 */       agg_unsafeRowAggBuffer_0.setLong(1, agg_value_17);
/* 347 */     } else {
/* 348 */       agg_unsafeRowAggBuffer_0.setNullAt(1);
/* 349 */     }
/* 350 */   }
...
```

### Why are the changes needed?

For high performance.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Existing tests.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala]: nit: Let's also update the comment here as well.

Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala]: ok, thanks.

**Comment by cloud-fan** [comment]
cc @rednaxelafx as well

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala
Comment (beliefer) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala]: @maropu  I don't understand why we need change `FilterExec` ?

Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala]: Just for sharing code to process predicates between aggregates and filters.

**Comment by maropu** [comment]
retest this please

**Comment by maropu** [comment]
retest this please

**Comment by maropu** [comment]
retest this please

**Comment by maropu** [comment]
retest this please

**Comment by kiszk** [comment]
Would it be possible to add benchmark result?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala
Comment (kiszk) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala]: nit: indentation problem?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala
Comment (kiszk) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala]: ditto

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala
Comment (kiszk) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala]: ditto

**Comment by maropu** [comment]
@kiszk I added the performance numbers in the PR description. I think the codegen for hash-aggregates can have performance gains in most queries. But, aggregate filters (recently merged in the master) forcibly disable the codegen. So, I think this fix has a good effect on performance.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala
Comment (kiszk) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala]: Can we refactor this and [that](https://github.com/apache/spark/pull/27019/files#diff-2eb948516b5beaeb746aadac27fbd5b4R997) to use one common function?

Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala]: Yea, I'll update.

**Comment by maropu** [comment]
retest this please

**Review by kiszk — APPROVED** [review]

**Comment by kiszk** [comment]
cc @rednaxelafx @cloud-fan

**Comment by kiszk** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala
Comment (xuzikun2003) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala]: I don't understand why "while(false)" can take an effect here. Could you explain why it is needed here?

Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala]: NOTE in above code comment would be enough to explain why, right? It still executes only once, but be able to exit the specific code block instead of exiting the function/method in the middle of the code.

Comment (xuzikun2003) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala]: Got it. Thanks for the explanation.

Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala]: Thanks for the reply, @HeartSaVioR. Since I forgot to add the generated code, I've added it in the PR description.

**Comment by maropu** [comment]
retest this please

**Comment by maropu** [comment]
retest this please

**Comment by maropu** [comment]
retest this please

**Comment by maropu** [comment]
cc: @cloud-fan

**Comment by maropu** [comment]
retest this please

**Comment by maropu** [comment]
retest this please

File: sql/core/src/test/resources/sql-tests/inputs/group-by-filter.sql
Comment (beliefer) [inline sql/core/src/test/resources/sql-tests/inputs/group-by-filter.sql]: I can't understand this dimension config.
When `spark.sql.codegen.wholeStage=false` and `spark.sql.codegen.factoryMode`.
It means each operator will be compiled into single java and cannot fallback to interpreted ?

Comment (maropu) [reply sql/core/src/test/resources/sql-tests/inputs/group-by-filter.sql]: That means the plan interpreter is  used with the expression codegen (e.g., `GenerateUnsafeProjection`) enabled.

Comment (beliefer) [reply sql/core/src/test/resources/sql-tests/inputs/group-by-filter.sql]: Thanks. I learned it.

**Comment by maropu** [comment]
retest this please

**Comment by maropu** [comment]
retest this please

**Comment by maropu** [comment]
retest this please

**Comment by maropu** [comment]
retest this please

**Comment by HeartSaVioR** [comment]
Hi @maropu, one question from curiosity.

I guess you've already got +1 from @kiszk, so unless you haven't made non-trivial change afterwards why don't you just go merge? Is there something you'd like to get reviewed, or is there someone you want to get reviewed before merging? If you are waiting for anything it would be nice to describe what you're waiting for.

**Comment by kiszk** [comment]
I think that @mapopu is waiting for LGTM from other appropriate developers such as @cloud-fan or @rednaxelafx.

**Comment by maropu** [comment]
Thanks, @kiszk. Yea, since the aggregate class is frequently-used one, so I just think more reviewers would be better to check the changes. As you know, the two release voting is concurrently in progress and we have low review bandwidth now. So, let me oepn this PR for a while. Anyway, thanks for your attention, @HeartSaVioR .

**Comment by HeartSaVioR** [comment]
Yeah I've indicated you're triggering builds continuously so guessed you might be stuck on reviewing, but realized you've already got +1 so wondered. Thanks for the explanation!

**Comment by HyukjinKwon** [comment]
I think it's okay to wait for more review to stay safer. I thought it was pretty clear that this PR is waiting for more review for that purpose. That's what I do too sometimes when I want to be very sure - even when I get multiple approvals from multiple committers :-).

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by maropu** [comment]
Thanks for re-trigering the tests, @HyukjinKwon

**Comment by maropu** [comment]
retest this please

**Comment by maropu** [comment]
retest this please

**Comment by maropu** [comment]
retest this please

**Comment by maropu** [comment]
retest this please

**Comment by maropu** [comment]
kindly ping @cloud-fan

**Comment by beliefer** [comment]
@maropu I'm sorry! Because the change of https://github.com/apache/spark/pull/27428, each filter clause will be converted to be normal plan. So we not need this PR again.

**Comment by maropu** [comment]
retest this please

**Comment by maropu** [comment]
retest this please

**Comment by maropu** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HeartSaVioR** [comment]
@maropu Looks like this PR is not properly tracked. I've removed the Stale tag assuming you want to move forward, but wanted to check again, and curious how you handle this.

**Comment by maropu** [comment]
Thanks for taking care of it, @HeartSaVioR. If there are no reviewer who is against this PR and one has no more comment, I will merge this for the v3.2.0 release in a few weeks.

**Comment by maropu** [comment]
retest this please

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Sorry for the delay and thank you, @maropu .
Merged to master for Apache Spark 3.2.0.

Merry Christmas and Happy New Year!

**Comment by maropu** [comment]
Thanks for your review, Dongjoon ! Yea, you, too. Happy Merry Christmas!

**Pull request closed without merging**
